src/Ocelot/Multiplexer/MultiplexingMiddleware.cs

```diff
@@ -1,219 +1,265 @@
+        var mainResponse = await ProcessMainRoute(httpContext, downstreamRoutes[0]);
+        if (mainResponse == null)
+        {
+            return;
+        }
 
-                for (var i = 0; i < route.DownstreamRoute.Count; i++)
-                {
-                    var newHttpContext = Copy(httpContext);
+        var tasksList = await ProcessRoutesWithRouteKeys(httpContext, downstreamRoutes, routeKeysConfigs, mainResponse);
+        if (tasksList.Length == 0)
+        {
+            return;
+        }
 
-                    newHttpContext.Items
-                        .Add("RequestId", httpContext.Items["RequestId"]);
-                    newHttpContext.Items
-                        .SetIInternalConfiguration(httpContext.Items.IInternalConfiguration());
-                    newHttpContext.Items
-                        .UpsertTemplatePlaceholderNameAndValues(httpContext.Items.TemplatePlaceholderNameAndValues());
-                    newHttpContext.Items
-                        .UpsertDownstreamRoute(route.DownstreamRoute[i]);
+        await MapResponses(httpContext, route, mainResponse, tasksList.ToList());
+    }
 
-                    tasks[i] = Fire(newHttpContext, _next);
-                }
+    /// <summary>
+    /// Helper method to determine if only the first downstream route should be processed.
+    /// It is the case if the request is a websocket request or if there is only one downstream route.
+    /// </summary>
+    /// <param name="context">The http context.</param>
+    /// <param name="routes">The downstream routes.</param>
+    /// <returns>True if only the first downstream route should be processed.</returns>
+    private static bool ShouldProcessSingleRoute(HttpContext context, ICollection routes) => context.WebSockets.IsWebSocketRequest || routes.Count == 1;
+
+    /// <summary>
+    /// Processing a single downstream route (no route keys).
+    /// In that case, no need to make copies of the http context.
+    /// </summary>
+    /// <param name="context">The http context.</param>
+    /// <param name="route">The downstream route.</param>
+    /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+    protected virtual Task ProcessSingleRoute(HttpContext context, DownstreamRoute route)
+    {
+        context.Items.UpsertDownstreamRoute(route);
+        return _next.Invoke(context);
+    }
 
-                await Task.WhenAll(tasks);
+    /// <summary>
+    /// Processing the downstream routes (no route keys).
+    /// </summary>
+    /// <param name="context">The main http context.</param>
+    /// <param name="route">The route.</param>
+    private async Task ProcessRoutes(HttpContext context, Route route)
+    {
+        var tasks = route.DownstreamRoute.Select(downstreamRoute => ProcessRouteAsync(context, downstreamRoute, _next))
+            .ToArray();
+        var contexts = await Task.WhenAll(tasks);
+        await Map(context, route, [.. contexts]);
```

Comment:
You haven't to await here! Just return `Task`.

Reply:
OK

Reply:
are you sure...?
```csharp
return Task.WhenAll(tasks).ContinueWith(t => Map(context, route, t.Result), TaskContinuationOptions.OnlyOnRanToCompletion);
```
Is it really any better?

Reply:
I would propose keeping it like it is...